src/app/treelist-checkbox/treelist-checkbox.component.ts

```diff
@@ -0,0 +1,250 @@
+      if (parent === this.data.dataset.name) {
+        return;
+      }
+      let flag = 0;
+      this.findNodeById(this.data, parent)?.children?.forEach(child => {
+        if (this.allStudies.has(child.dataset.name)) {
+          flag += 1;
+        }
+      });
+      if (flag === parents.length) {
+        flag = 2;
+      } else if (flag !== 0 && flag < parents.length) {
+        flag = 1;
+      }
+      switch (flag) {
+        case 0: {
+          const elements = document.querySelectorAll(`[id='${parent}']`);
+          elements.forEach(element => {
+            (element as HTMLInputElement).checked = false;
+            (element as HTMLInputElement).indeterminate = false;
+          });
+          break;
+        }
+        case 1: {
+          const elements = document.querySelectorAll(`[id='${parent}']`);
+          elements.forEach(element => {
+            (element as HTMLInputElement).checked = true;
+            (element as HTMLInputElement).indeterminate = true;
+          });
+          break;
+        }
+        case 2: {
+          const elements = document.querySelectorAll(`[id='${parent}']`);
+          elements.forEach(element => {
+            (element as HTMLInputElement).checked = true;
+            (element as HTMLInputElement).indeterminate = false;
+          });
+          break;
+        }
+      }
+    });
+  }
+
+  public updateFilterStates(data: DatasetNode): void {
+    let checkState = 0;
+    if (data.children.length > 0) {
+      data.children.forEach(child => {
+        if (child.children.length > 0) {
+          this.updateFilterStates(child);
+        }
+      });
+    }
+
+    const children = this.getAllChildren(data.children, new Set());
+    this.getAllChildren(data.children, new Set()).forEach(child => {
+      if (this.allStudies.has(child)) {
+        checkState = 1;
+      }
+    });
+    let flag = 1;
+    children.forEach(child => {
+      if (!this.allStudies.has(child)) {
+        flag = 0;
+      }
+    });
+    if (flag) {
+      checkState = 2;
+    }
+    switch (checkState) {
+      case 0: {
+        const elements = document.querySelectorAll(`[id='${data.dataset.name}']`);
+        elements.forEach(element => {
+          (element as HTMLInputElement).checked = false;
+          (element as HTMLInputElement).indeterminate = false;
+        });
+        break;
+      }
+      case 1: {
+        const elements = document.querySelectorAll(`[id='${data.dataset.name}']`);
+        elements.forEach(element => {
+          (element as HTMLInputElement).checked = true;
+          (element as HTMLInputElement).indeterminate = true;
+        });
+        break;
+      }
+      case 2: {
+        const elements = document.querySelectorAll(`[id='${data.dataset.name}']`);
+        elements.forEach(element => {
+          (element as HTMLInputElement).checked = true;
+          (element as HTMLInputElement).indeterminate = false;
+        });
+        break;
+      }
+    }
```

Comment:
Repeating code - you could reduce it using ternary operators when assigning the `checked` and `indeterminate` values of the checkbox element.